### PR TITLE
SW-2740 migration script

### PR DESCRIPTION
**Pull request by artem7s**
Usage:
```bash
yarn migrate --shops=1,125 [--dry-run]
```

- transforms and migrate analytics email templates and sending logic into notification DB
  - deleted sending logic (when `deletedAt` is defined) is not getting migrated
  - migrated templates have `analyticsId` field 
  - migrated widgets have `analyticsId` and `analyticsTemplateId` fields
  - migrated sending logics have `analyticsId` and `analyticsTemplateId` fields
  - `emailFromAddressIsValidated` is true when sending logic is `live` or when email verified in `email_verification` table or when it is whitelisted
  - `settings` fields have special processing to prevent them being empty arrays instead of empty objects
  - transforms `copyToEmails` from string to array (splits like FE does it)
  - transforms `emailSendTo` values 
- sets unique index for `analyticsId` fields in `emailtemplates` and `sendinglogics` collections
- checks if shop already has any data
- import wrapped unto transaction
- configuration is read from `.env` file

image

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,2 +1,7 @@
 # notifications-migration
 Script for migrating notifications from Analytics DB to Notifications DB
+
+## Usage
+```bash
+yarn migrage --shops=1,175 [--dry-run]
```
Comment (andrey7s) [inline README.md:6]: typo

File: src/index.js [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+  const liveSendingLogicsCount = analyticsSendingLogics.filter((sl) => sl.status === 'live').length;
+
+  console.log(
+    `Found ${chalk.yellow(analyticsSendingLogics.length)} sending logics (${chalk.green(liveSendingLogicsCount)} live)`
+  );
+
+  const notificationsDB = await getNotificationsMongoDatabase();
+
+  if ((await getEmailTemplatesCollection(notificationsDB).count({ shopId })) > 0) {
+    console.error(chalk.red('Found existing email templates in Notifications DB!'));
```
Comment (andrey7s) [inline src/index.js:73]: What's the plan for such cases?

File: src/index.js [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+    const { shops, dryRun } = argv;
+    const shopIds = shops.split(',');
+
+    await migrateCommand({ shopIds, dryRun });
+  }
+
+  process.exit();
+}
+
+main().catch((err) => {
```
Comment (andrey7s) [inline src/index.js:110]: Would it be safer to run `main()` only if `index.js` is executed directly and not just imported?

File: src/notifications/import.js [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+  const db = await getNotificationsMongoDatabase();
+  const templatesInsertResult = await getEmailTemplatesCollection(db).insertMany(templates, { session });
+
+  const mappingAnalyticsIdToMongoId = getIdsMapping(templates, templatesInsertResult.insertedIds);
+  const sendingLogicsWithTemplateId = sendingLogics.map((logic) => {
+    const { analyticsTemplateId } = logic;
+    const templateId = analyticsTemplateId ? mappingAnalyticsIdToMongoId[analyticsTemplateId] : null;
+
+    if (analyticsTemplateId && !templateId) {
+      console.warn(`Can't fine template's Mongo ID for Analytics ID ${analyticsTemplateId}`);
```
Comment (andrey7s) [inline src/notifications/import.js:60]: typo "Can't find ..."

File: src/notifications/import.js [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+  return {
+    templatesInsertedCount: templatesInsertResult.insertedCount,
+    sendingLogicsInsertedCount: sendingLogicsInsertResult.insertedCount,
+  };
+}
+
+function getIdsMapping(templates, insertedIds) {
+  const result = {};
+
+  templates.forEach((template, index) => (result[template.analyticsTemplateId] = insertedIds[index].toString()));
```
Comment (andrey7s) [inline src/notifications/import.js:79]: Is it really safe to rely just on the index and the same order?

File: README.md [hunk 6]
```diff
@@ -1,2 +1,7 @@
 # notifications-migration
```
Comment (greensender) [inline README.md:1]: why in different repo and not in a branch tagged in notifications service?

File: src/transformers.js [hunk 7]
```diff
@@ -0,0 +1,133 @@
...
+  return (strEmails || '')
+    .split(/,| /)
+    .map((x) => x.trim())
+    .filter((x) => x);
+}
+
+const sendToMapping = {
+  customers: 'customers',
+  shop: 'shop',
+  shop_and_customers: 'shopAndCustomers',
```
Comment (andrey7s) [inline src/transformers.js:107]: It's `shop_and_customers`

File: src/transformers.js [hunk 8]
```diff
@@ -0,0 +1,133 @@
...
+    createdAt,
+    updatedAt,
+  };
+}
+
+function transformAnalyticsTemplateWidget(widget) {
+  const { id: analyticsWidgetId, notificationTemplateId: analyticsTemplateId, type, settings } = widget;
+
+  return {
+    _id: ObjectId(),
```
Comment (andrey7s) [inline src/transformers.js:36]: I'm not sure if `_id` is necessary for widgets.

File: src/analytics/sendingLogics.js [hunk 9]
```diff
@@ -0,0 +1,16 @@
+const { getConnection, IsNull } = require('typeorm');
+
+async function getShopSendingLogics(shopId) {
+  return getConnection()
+    .getRepository('SendingLogic')
+    .find({
+      where: {
+        shopId,
+        deletedAt: IsNull(),
```
Comment (greensender) [inline src/analytics/sendingLogics.js:9]: may be if we keep the name `getShopSendingLogics` but filter by deletedAt, it's better to pass extra arguments, like {deletedAt: IsNull()}, or rename to getNotDeletedSendingLogics ?

File: src/transformers.js [hunk 10]
```diff
@@ -0,0 +1,133 @@
...
+    analyticsWidgetId,
```
Comment (andrey7s) [inline src/transformers.js:37]: I'm not sure if `widgetId` is important.

File: src/transformers.js [hunk 11]
```diff
@@ -0,0 +1,133 @@
...
+
+  return {
+    analyticsTemplateId,
+    shopId,
+    localizationId,
+    name,
+    type,
+    content,
+    widgets,
+    settings: transformSettings(settings),
```
Comment (greensender) [inline src/transformers.js:26]: shouldn't the settings be null for html templates?

File: src/analytics/entities/Template.js [hunk 12]
```diff
@@ -0,0 +1,29 @@
+const { EntitySchema } = require('typeorm');
+
+function settingsFromTransformer(strData) {
```
Comment (greensender) [inline src/analytics/entities/Template.js:3]: could be exported?

File: src/index.js [hunk 13]
```diff
@@ -0,0 +1,113 @@
...
+
+  for (const shopId of shopIds) {
+    await migrateShop({ shopId, dryRun });
+  }
+}
+
+async function migrateShop({ shopId, dryRun }) {
+  const shop = await getShop(shopId);
+
+  if (!shop) {
```
Comment (greensender) [inline src/index.js:52]: do we need to check that the shop being migrated has migrationMode == migrating?

File: README.md (hunk 6 above)
Comment (artem7s) [reply README.md:1]: Main reason it's not just a script but a small program with bunch of deps which not related to notification service

File: src/index.js (hunk 2 above)
Comment (artem7s) [reply src/index.js:73]: It shouldn't be the case. It means that most likely the data has been already migrated.

File: src/index.js (hunk 3 above)
Comment (artem7s) [reply src/index.js:110]: ? `index.js` isn't getting imported anywhere

**Comment by andrey7s** [comment]
The main question: how do plan to run the migration for the production database?

File: src/transformers.js (hunk 8 above)
Comment (artem7s) [reply src/transformers.js:36]: Me neither, but if I don't provide `_id` fetching email template fails with a message that `id` field should be present in response.

File: src/transformers.js (hunk 7 above)
Comment (artem7s) [reply src/transformers.js:107]: The value?

File: src/index.js (hunk 13 above)
Comment (andrey7s) [reply src/index.js:52]: Actually that's a good point.

Comment (artem7s) [reply src/index.js:52]: Yeah, makes sense

File: src/notifications/import.js (hunk 5 above)
Comment (artem7s) [reply src/notifications/import.js:79]: They claim it is:
> The index of each ID in the map corresponds to each document’s position in the bulk operation.

**Pull request merged**